clippy_lints/src/redundant_clone.rs

```diff
@@ -70,317 +68,724 @@ impl<'tcx> LateLintPass<'tcx> for RedundantClone {
+                });
+                if let TerminatorKind::Call { args, .. } = &data.terminator().kind
+                    && let [Operand::Copy(arg) | Operand::Move(arg), ..] = &**args
+                    && let Some(Some(targets)) = ref_target_results.get().targets.get(&arg.local)
+                {
+                    Some((block, (arg.local, targets.clone())))
+                } else {
+                    None
+                }
+            })
+            .collect::<FxHashMap<_, _>>();
+        if clone_sources.is_empty() {
+            return;
+        }
 
-            if terminator.source_info.span.from_expansion() {
-                continue;
+        let mut results = CloneAnalysis::new(cx, body, &raw_ptr_taken, clone_sources)
+            .into_engine(cx.tcx, body)
+            .iterate_to_fixpoint()
+            .into_results_cursor(body);
+
+        let mut candidates = Candidates::default();
+        for (block, data) in body.basic_blocks.iter_enumerated() {
+            match data.terminator().kind {
+                TerminatorKind::Return => {
+                    results.seek_to_block_end(block);
+                    results.apply_custom_effect(|a, state| {
+                        a.read(state, Local::from_u32(0));
+                    });
+                    candidates.join(&results.get().candidates);
+                },
+                TerminatorKind::Terminate | TerminatorKind::Resume => {
+                    results.seek_to_block_end(block);
+                    candidates.join(&results.get().candidates);
+                },
+                _ => {},
             }
+        }
 
-            // Give up on loops
-            if terminator.successors().any(|s| s == bb) {
-                continue;
+        for (&block, &candidate) in &candidates.0 {
+            if candidate == Candidate::Lint {
+                let term = body.basic_blocks[block].terminator();
+                let node = body.source_scopes[term.source_info.scope]
+                    .local_data
+                    .as_ref()
+                    .assert_crate_local()
+                    .lint_root;
+                span_lint_hir(cx, REDUNDANT_CLONE, node, term.source_info.span, "redundant clone");
             }
+        }
+    }
+}
 
-            let (fn_def_id, arg, arg_ty, clone_ret) =
-                unwrap_or_continue!(is_call_with_ref_arg(cx, mir, &terminator.kind));
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum Candidate {
+    Lint,
+    NoLint,
+}
 
-            let from_borrow = match_def_path(cx, fn_def_id, &paths::CLONE_TRAIT_METHOD)
-                || match_def_path(cx, fn_def_id, &paths::TO_OWNED_METHOD)
-                || (match_def_path(cx, fn_def_id, &paths::TO_STRING_METHOD)
-                    && is_type_lang_item(cx, arg_ty, LangItem::String));
+#[derive(Debug, Clone)]
+struct Value {
+    last_modified: Option<BasicBlock>,
+    lint_unread: HybridBitSet<BasicBlock>,
```

Comment:
This field looks like it attempts to model a liveness analysis. Can it be replaced by `MayleLiveLocals` that we already have in store?

Reply:
This is tracking if a read has occurred since the paired value was first modified.